### PR TITLE
feat(idempotency): add support to custom serialization/deserialization on idempotency decorator

### LINKED ISSUE #2886: Feature request: Support of custom serizialization/deserisialization (object -> dict) on idempotency decorator 'idempotent_function'
### Use case

I use Pydantic as a representation of my DTO. As such, I receive and return pydantic objects as part of my internal function calls.
Currently, the idempotency decorator supports returning dictionary only (this was also discussed here #2164 ).

I would like to be able bring my own serialization logic in order to support custom DTO implementations.

### Solution/User Experience

User will be able to add a serailizer/deserializer in order to control how it is done using simple function parameters:
``` python
def idempotent_function(
    function: Optional[AnyCallableT] = None,
    *,
    data_keyword_argument: str,
    persistence_store: BasePersistenceLayer,
    config: Optional[IdempotencyConfig] = None,
    input_serializer: Optional[Callable[[Any], Dict]] = None,
    output_serializer: Optional[Callable[[Any], Dict]] = None,
    output_deserializer: Optional[Callable[[Dict], Any]] = None,
) -> Any:
```
For example 
``` python
  class PaymentOutput(BaseModel):
      customer_id: str
      transaction_id: str

  @idempotent_function(
      data_keyword_argument="payment",
      persistence_store=persistence_layer,
      config=config,
      input_serializer=lambda x: x.dict(),
      output_serializer=lambda x: x.dict(),
      output_deserializer=PaymentOutput.parse_obj,
  )
) -> Any:
```

As there is currently an existing implementation of input serialization (object -> dict), not supplying any input_serializer will use the existing "best effort" implementation.

### Alternative solutions

User will be able to add a serailizer/deserializer in order to control how it is done using the IdempotencyConfig:
``` python

class IdempotencyConfig:
    def __init__(
        self,
        event_key_jmespath: str = "",
        payload_validation_jmespath: str = "",
        jmespath_options: Optional[Dict] = None,
        raise_on_no_idempotency_key: bool = False,
        expires_after_seconds: int = 60 * 60,  # 1 hour default
        use_local_cache: bool = False,
        local_cache_max_items: int = 256,
        hash_function: str = "md5",
        lambda_context: Optional[LambdaContext] = None,
        input_serializer: Optional[Callable[[Any], Dict]] = None,
        output_serializer: Optional[Callable[[Any], Dict]] = None,
        output_deserializer: Optional[Callable[[Dict], Any]] = None,
    ):
```

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by aradyaron**
**Issue number:** #2886 
 
### Changes

* New parameter in idempotent function handler : output_serializer
* New base class for dict serializer, this allows further internal changes without breaking interfaces
* CustomDict implementation for serializer
* Pydantic implementation for serializer
* Tests for both serializaer types
* Examples for both serializer types
* Documentation regarding the new output_serializer parameter
### User experience

* Users can now add output serializer on idempotent function decorator in order to use custom Dto types on their functions

```python
from aws_lambda_powertools.utilities.idempotency import (
    DynamoDBPersistenceLayer,
    IdempotencyConfig,
    idempotent_function,
)
from aws_lambda_powertools.utilities.idempotency.serialization.pydantic import PydanticSerializer
from aws_lambda_powertools.utilities.parser import BaseModel
from aws_lambda_powertools.utilities.typing import LambdaContext

dynamodb = DynamoDBPersistenceLayer(table_name="IdempotencyTable")
config = IdempotencyConfig(event_key_jmespath="order_id")  # see Choosing a payload subset section


class OrderItem(BaseModel):
    sku: str
    description: str


class Order(BaseModel):
    item: OrderItem
    order_id: int


class OrderOutput(BaseModel):
    order_id: int


@idempotent_function(
    data_keyword_argument="order",
    config=config,
    persistence_store=dynamodb,
    output_serializer=PydanticSerializer,
)
# order output is deduced from return type
def deduced_order_output_serializer(order: Order) -> OrderOutput:
    return OrderOutput(order_id=order.order_id)


def lambda_handler(event: dict, context: LambdaContext):
    config.register_lambda_context(context)  # see Lambda timeouts section
    order_item = OrderItem(sku="fake", description="sample")
    order = Order(item=order_item, order_id=1)

    # `order` parameter must be called as a keyword argument to work
    deduced_order_output_serializer(order=order)
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @aradyaron! Thanks so much for sending this PR, it's absolutely awesome to add this support in the Idempotency utility without breaking changes.

We will start reviewing this PR by tomorrow and hope to merge it as soon as possible.

:rocket:

**Comment by rubenfonseca** [comment]
Looking at this now

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 1]
```diff
@@ -64,13 +68,17 @@ def __init__(
         config: IdempotencyConfig
             Idempotency Configuration
         persistence_store : BasePersistenceLayer
-            Instance of persistence layer to store idempotency records
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py:67]: Can we get this line back again? :)

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 2]
```diff
@@ -226,7 +236,8 @@ def _get_function_response(self):
 
         else:
             try:
-                self.persistence_store.save_success(data=self.data, result=response)
+                saved_response: dict = self.output_serializer.to_dict(response) if self.output_serializer else response
+                self.persistence_store.save_success(data=self.data, result=saved_response)
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py]:
```suggestion
                serialized_response: dict = self.output_serializer.to_dict(response) if self.output_serializer else response
                self.persistence_store.save_success(data=self.data, result=serialized_response)
```

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 3]
```diff
@@ -43,7 +44,6 @@ def idempotent(
         Instance of BasePersistenceLayer to store data
     config: IdempotencyConfig
         Configuration
-
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]: Can we get this empty line back so we can separate the parameters documentation from examples?

File: aws_lambda_powertools/utilities/idempotency/idempotency.py [hunk 4]
```diff
@@ -141,12 +145,12 @@ def decorate(*args, **kwargs):
             )
 
         payload = kwargs.get(data_keyword_argument)
-
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/idempotency.py]: Can we get this empty line back?

File: aws_lambda_powertools/utilities/idempotency/__init__.py [hunk 5]
```diff
@@ -8,7 +8,17 @@
 from aws_lambda_powertools.utilities.idempotency.persistence.dynamodb import (
     DynamoDBPersistenceLayer,
 )
+from aws_lambda_powertools.utilities.idempotency.serialization.custom_dict import CustomDictSerializer
+from aws_lambda_powertools.utilities.idempotency.serialization.pydantic import PydanticSerializer
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/__init__.py]: Unfortunately we can't export this globally: we should be able to use the idempotency utility without having Pydantic installed.

Suggestion: don't export globally here, and just keep it inside the serialization folder. This way, when people want to use it, they have to:

`import PydanticSerializer from aws_lambda_powertools.utilities.serialization`

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
First pass, haven't looked into the tests and docs yet. Great job!

Here's a couple of comments:
- We need to change things a little bit, since Pydantic dependency is optional
- I would like you to consider using [Tell Don't Ask](https://martinfowler.com/bliki/TellDontAsk.html):

Create a new NoOp serializer:

```python
class NoOpSerializer(BaseSerializer):
    def to_dict(self, data: Any) -> Dict:
        return data

    def from_dict(self, data: Dict) -> Any:
        return data
```

Use it as the default parameter on the decorator:

`output_serializer: BaseDictSerializer = NoOpSerializer()`

And this way you can remove all the`if self.output_serializer` in the code.

**Comment by rubenfonseca** [comment]
Another thing that I'm concerned is that we currently always go through JSON, which seems innefective. Take a look at this example:

image

If your response is actually a Pydantic model, we are transforming the model into a dictionary, just to parse it again with Pydantic in order to save it. What do you feel about this?

**Comment by aradyaron** [comment]
> payload = kwargs.get(data_keyword_argument)

@rubenfonseca You are indeed correct.
This has been taken into consideration, but due to how the idmepotency is currently written, this kind of change requires a bigger refactor.
By using the serializer class, and inheritance for specific cases (e.g pydantic) this will allow further optimization of this flow

File: aws_lambda_powertools/utilities/idempotency/__init__.py (hunk 5 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/__init__.py]: Hey, thanks for noticing!
removed on __init__.py and updated examples

File: aws_lambda_powertools/utilities/idempotency/idempotency.py (hunk 4 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/idempotency.py]: done

File: aws_lambda_powertools/utilities/idempotency/idempotency.py (hunk 3 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/idempotency.py]: done

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 2 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/base.py]: changed to serializer :)

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 1 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/base.py:67]: Added back, thanks for noticing!

**Comment by aradyaron** [comment]
> Another thing that I'm concerned is that we currently always go through JSON, which seems innefective. Take a look at this example:
> 
> image
> 
> If your response is actually a Pydantic model, we are transforming the model into a dictionary, just to parse it again with Pydantic in order to save it. What do you feel about this?

@rubenfonseca You are indeed correct.
This has been taken into consideration, but due to how the idmepotency is currently written, this kind of change requires a bigger refactor.
By using the serializer class, and inheritance for specific cases (e.g pydantic) this will allow further optimization of this flow upon refactoring, without breaking the interface

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 6]
```diff
@@ -51,6 +57,7 @@ def __init__(
         function_payload: Any,
         config: IdempotencyConfig,
         persistence_store: BasePersistenceLayer,
+        output_serializer: Optional[BaseDictSerializer] = None,
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py]: I think you meant:

```suggestion
        output_serializer: BaseDictSerializer = NoOpSerializer(),
```

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 7]
```diff
@@ -65,12 +72,17 @@ def __init__(
             Idempotency Configuration
         persistence_store : BasePersistenceLayer
             Instance of persistence layer to store idempotency records
+        output_serializer: Optional[BaseDictSerializer]
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py]:
```suggestion
        output_serializer: BaseDictSerializer
```

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 8]
```diff
@@ -65,12 +72,17 @@ def __init__(
...
+            Serializer to transform the data to and from a dictionary.
+            If not supplied, no serialization is done
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py]:
```suggestion
            If not supplied, no serialization is done via the NoOpSerializer
```

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
I think this is going in the right direction!! Thank you so much for applying the changes!

What do you think of the suggestion? Do you see any problem with it, making the parameter not optional with a default? If it's ok, can you please change it in the `idempotent_function` too?

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 6 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/base.py]: Once I try this change I receive the 'make pr' error
B008 Do not perform function call `NoOpSerializer` in argument defaults.

1. Should I still perform this change?
2. If I should, how can I bypass/ignore this error?

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 7 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/base.py]: Currently the class allows to not pass 'output_serializer' parameter.
Is this change correct?

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+"""
+Serialization for supporting idempotency
+"""
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+logger = logging.getLogger(__name__)
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/base.py]: I don't see the logger being used here, can we remove it?

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class BaseDictSerializer(ABC):
+    """
+    Abstract Base Class for Idempotency serialization layer, supporting dict operations.
+    This interface not be inherited by end user implementation.
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/base.py]: Can you rephrase this sentence? I'm not sure what you mean here. maybe we're missing a "should"?

```suggestion
    This interface should not be inherited by end user implementation.
```

File: aws_lambda_powertools/utilities/idempotency/serialization/custom_dict.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class CustomDictSerializer(BaseDictSerializer):
+    def __init__(self, to_dict: Callable[[Any], Dict], from_dict: Callable[[Dict], Any]):
+        """
+        Parameters
+        ----------
+        to_dict: Callable[[Any], Dict]
+            A function capable of transforming the saved data object representation into a dictionary
+        from_dict: str
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/custom_dict.py]: I believe this should be

```suggestion
        from_dict: Callbable[[Dict], Any]
```

File: aws_lambda_powertools/utilities/idempotency/serialization/no_op.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+from aws_lambda_powertools.utilities.idempotency.serialization.base import BaseDictSerializer
+
+
+class NoOpSerializer(BaseDictSerializer):
+    def __init__(self):
+        """
+        Parameters
+        ----------
+        default serializer, does not transform data
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/no_op.py]:
```suggestion
        Default serializer, does not transform data
```

File: aws_lambda_powertools/utilities/idempotency/serialization/pydantic.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+Model = BaseModel
+
+
+class PydanticSerializer(BaseDictSerializer):
+    def __init__(self, model: Type[Model]):
+        """
+        Parameters
+        ----------
+        model: Model
+            A model of the type to transform
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/pydantic.py]:
```suggestion
            A Pydantic model of the type to transform
```

File: examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    item: OrderItem
+    order_id: int
+
+
+@dataclass
+class OrderOutput:
+    order_id: int
+
+
+order_output_serializer: CustomDictSerializer = CustomDictSerializer(
```
Comment (rubenfonseca) [inline examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py]: No need to repeat the type here:

```suggestion
order_output_serializer = CustomDictSerializer(
```

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 15]
```diff
@@ -65,12 +72,17 @@ def __init__(
...
+            Serializer to transform the data to and from a dictionary.
+            If not supplied, no serialization is done via the NoOpSerializer
         function_args: Optional[Tuple]
             Function arguments
         function_kwargs: Optional[Dict]
             Function keyword arguments
+
         """
         self.function = function
+        self.output_serializer = output_serializer or NoOpSerializer()
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/base.py:84]: This is great, so I'm removing all my suggestions to pass it in the function declaration.

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 7 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/idempotency/base.py]: I reconsidered :D Your current code looks great, so ignore this suggestion.

File: aws_lambda_powertools/utilities/idempotency/base.py (hunk 6 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/idempotency/base.py]: Sorry, I reconsidered. Your code looks good, leave it as it is!

File: examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    order_id: int
+
+
+@idempotent_function(
+    data_keyword_argument="order",
+    config=config,
+    persistence_store=dynamodb,
+    output_serializer=PydanticSerializer(model=OrderOutput),
+)
+def process_order(order: Order) -> OrderOutput:
```
Comment (rubenfonseca) [inline examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py]: Thank you for creating this sample! I'm not entirely happy that as a user I have to repeat the `OrderOutput` twice, when the information may be inferred from the signature. Do you feel like trying that?

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+Serialization for supporting idempotency
+"""
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+logger = logging.getLogger(__name__)
+
+
+class BaseDictSerializer(ABC):
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/idempotency/serialization/base.py]: What do you think calling this `BaseSerializer` instead? It's more generic.

**Review by rubenfonseca — CHANGES_REQUESTED** [review]
Great job! I've checked the initial docs and tests now, and they look good! Left a couple of small comments to think about :) Almost there!

File: examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py (hunk 16 above)
Comment (aradyaron) [reply examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py]: This sounds like a nice idea, I'll investigate and update

**Comment by aradyaron** [comment]
> > Another thing that I'm concerned is that we currently always go through JSON, which seems innefective. Take a look at this example:
> > image
> > If your response is actually a Pydantic model, we are transforming the model into a dictionary, just to parse it again with Pydantic in order to save it. What do you feel about this?
> 
> @rubenfonseca You are indeed correct. This has been taken into consideration, but due to how the idmepotency is currently written, this kind of change requires a bigger refactor. By using the serializer class, and inheritance for specific cases (e.g pydantic) this will allow further optimization of this flow upon refactoring, without breaking the interface

@rubenfonseca I have further checked and It's indeed possible to add this improvement, but adding it WILL break the persistence interface as can be seen here https://github.com/aws-powertools/powertools-lambda-python/blob/53650982c51f54cf448100a819aadb1ea5a23489/aws_lambda_powertools/utilities/idempotency/persistence/base.py#L314
The result type need to be updated to string instead of dict,
unless the Serializer interface will be something like 
* to_dict(Any) -> dict
* deserialize(string) -> Any

Which I think is a wrong non symmetric interface 
WDYT?

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 17 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py:11]: I also thought about it.
This comes back to the question if there will be a refactor to 'direct' serialization (see  https://github.com/aws-powertools/powertools-lambda-python/pull/2951#issuecomment-1693242636)
Because if we would like a refactor, and you already use *BaseSerializer*, you cannot addd an intermediate class of * BaseDictSerializer* and only change the internal implementation

Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py]: updated to BaseIdempotencySerializer

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 10 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py]: removed the line alltogehter

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 9 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py]: done

File: examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py (hunk 16 above)
Comment (aradyaron) [reply examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py]: Hey @rubenfonseca, I've added this behavior and developed this flow in a separate branch (in order to keep things clean)
please take a look here
https://github.com/aradyaron/powertools-lambda-python/pull/2 and let me know your opinion (also added a bonus of dataclass serializer :) )

if you prefer that I'll merge and then we'll discuss in this Pr please let me know.
Please note that additional examples and docs is needed, which I'll write once it's decided to add this functionality as well

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class BaseIdempotencySerializer(ABC):
+    """
+    Abstract Base Class for Idempotency serialization layer, supporting dict operations.
+    """
+
+    @abstractmethod
+    def to_dict(self, data: Any) -> Dict:
+        pass
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/idempotency/serialization/base.py]: Raising `NotImplementedError` is better because we increase the code coverage and follow the Python standards.

Reference: https://docs.python.org/3/library/exceptions.html#NotImplementedError

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+    @abstractmethod
+    def from_dict(self, data: Dict) -> Any:
+        pass
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/idempotency/serialization/base.py:47]: Same here.

File: examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    to_dict=lambda x: x.asdict(),
+    from_dict=lambda x: OrderOutput(**x),
```
Comment (leandrodamascena) [inline examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py]: What do you think about creating this function as a standalone? I mean, sometimes the `lambda` syntax makes it difficult to understand the proposal and increases the cognitive load.

Maybe this:

```python
def custom_to_dict(x: Any) -> Dict:
    return dataclasses.asdict(x)

def custom_from_dict(x: Dict) -> Any:
    return OrderOutput(**x)

order_output_serializer = CustomDictSerializer(
    to_dict_func=custom_to_dict,
    from_dict_func=custom_from_dict,
) 
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @aradyaron, awesome job adding support for custom serialization/deserialization in our Idempodentcy utility!

I made some few comments to change some small things.

Thank you.

**Comment by rubenfonseca** [comment]
> > > Another thing that I'm concerned is that we currently always go through JSON, which seems innefective. Take a look at this example:
> > > image
> > > If your response is actually a Pydantic model, we are transforming the model into a dictionary, just to parse it again with Pydantic in order to save it. What do you feel about this?
> > 
> > 
> > @rubenfonseca You are indeed correct. This has been taken into consideration, but due to how the idmepotency is currently written, this kind of change requires a bigger refactor. By using the serializer class, and inheritance for specific cases (e.g pydantic) this will allow further optimization of this flow upon refactoring, without breaking the interface
> 
> @rubenfonseca I have further checked and It's indeed possible to add this improvement, but adding it WILL break the persistence interface as can be seen here
> 
> https://github.com/aws-powertools/powertools-lambda-python/blob/53650982c51f54cf448100a819aadb1ea5a23489/aws_lambda_powertools/utilities/idempotency/persistence/base.py#L314
> 
> 
> The result type need to be updated to string instead of dict,
> unless the Serializer interface will be something like
> * to_dict(Any) -> dict
> * deserialize(string) -> Any
> 
> Which I think is a wrong non symmetric interface WDYT?

You are right, this would be a breaking change. We might as well keep the non-optimal behavior as it is, and do a re-write when we release Powertools v3 to optimize it. Sounds good?

File: examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py (hunk 16 above)
Comment (rubenfonseca) [reply examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py]: I've checked the branch and I think this will be a much better experience. I also liked that you made it generic enough to support other data types! Please merge the code, and we can review it here (just small naming things here and there). Awesome!

**Review by rubenfonseca — DISMISSED** [review]
- pending merge of the output serializer model detection code

File: examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py (hunk 20 above)
Comment (aradyaron) [reply examples/idempotency/src/working_with_idempotent_function_dataclass_output_serializer.py]: Nice!
changed

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 19 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py:47]: added

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 18 above)
Comment (aradyaron) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py]: Added :)

File: examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py (hunk 16 above)
Comment (aradyaron) [reply examples/idempotency/src/working_with_idempotent_function_pydantic_output_serializer.py]: Hey @rubenfonseca, I've merged the Pr.
later on I'll also add additional examples + docs

File: aws_lambda_powertools/utilities/idempotency/serialization/base.py (hunk 18 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/idempotency/serialization/base.py]: Thanks!

**Comment by aradyaron** [comment]
> * pending merge of the output serializer model detection code

Hey @rubenfonseca, code is merged

**Review by leandrodamascena — DISMISSED** [review]

**Comment by leandrodamascena** [comment]
I'm working with @sthulb to review the documentation and merge this PR. :rocket:

File: docs/utilities/idempotency.md [hunk 21]
```diff
@@ -152,6 +152,36 @@ When using `idempotent_function`, you must tell us which keyword parameter in yo
     --8<-- "examples/idempotency/src/working_with_idempotent_function_pydantic.py"
     ```
 
+#### Output Serialization
```
Comment (sthulb) [inline docs/utilities/idempotency.md]: The 's' should be lower case here.

Comment (sthulb) [reply docs/utilities/idempotency.md]: Overall, I think the docs are quite light on what functionality this feature provides and I think the text around the examples could be more detailed to explain how the examples work.

Comment (sthulb) [reply docs/utilities/idempotency.md]: I think comments in the code samples would also be appreciated by new users

**Review by sthulb — CHANGES_REQUESTED** [review]

File: docs/utilities/idempotency.md (hunk 21 above)
Comment (leandrodamascena) [reply docs/utilities/idempotency.md]: Hey @sthulb! Thanks for this feedback. I made some changes, can you check, please?

Comment (sthulb) [reply docs/utilities/idempotency.md]: Much better :)

**Review by sthulb — APPROVED** [review]

**Review by leandrodamascena — APPROVED** [review]

**Review by leandrodamascena — APPROVED** [review]
